File: use-cases/merchant-fulfillment/code/java/src/main/java/lambda/CreateShipmentHandler.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.google.gson.Gson;
import io.swagger.client.api.MerchantFulfillmentApi;
import io.swagger.client.model.mfn.CreateShipmentRequest;
import io.swagger.client.model.mfn.CreateShipmentResponse;
import io.swagger.client.model.mfn.Label;
import io.swagger.client.model.mfn.ShipmentRequestDetails;
import lambda.utils.MfnLambdaInput;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.PutItemRequest;

import java.util.HashMap;
import java.util.Map;

import static lambda.utils.ApiUtils.getMFNApi;
import static lambda.utils.Constants.SHIPMENTS_TABLE_HASH_KEY_NAME;
import static lambda.utils.Constants.SHIPMENTS_TABLE_NAME_ENV_VARIABLE;
import static lambda.utils.Constants.SHIPMENTS_TABLE_SHIPMENT_ID_ATTRIBUTE_NAME;
import static lambda.utils.MfnUtils.getDefaultShippingServiceOptions;
import static lambda.utils.MfnUtils.getItemList;

public class CreateShipmentHandler implements RequestHandler<MfnLambdaInput, Label> {

    public Label handleRequest(MfnLambdaInput input, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("CreateShipment Lambda input: " + new Gson().toJson(input));

        try {
            //Create shipment for the selected shipping service
            CreateShipmentRequest request = new CreateShipmentRequest()
                    .shipmentRequestDetails(new ShipmentRequestDetails()
                            .amazonOrderId(input.getOrderId())
                            .itemList(getItemList(input.getMfnOrder().getOrderItems()))
                            .shipFromAddress(input.getMfnOrder().getShipFromAddress())
                            .packageDimensions(input.getMfnOrder().getPackageDimensions())
                            .weight(input.getMfnOrder().getWeight())
                            .shippingServiceOptions(getDefaultShippingServiceOptions()))
                    .shippingServiceId(input.getMfnOrder().getPreferredShippingService().getShippingServiceId())
                    .shippingServiceOfferId(input.getMfnOrder().getPreferredShippingService().getShippingServiceOfferId());

            logger.log("Merchant Fulfillment API - CreateShipment request: " + new Gson().toJson(request));

            MerchantFulfillmentApi mfnApi = getMFNApi(input.getRegionCode(), input.getRefreshToken());
            CreateShipmentResponse response = mfnApi.createShipment(request);

            //Store ShipmentId in DynamoDB
            //Update this section to match your product's logic
            String shipmentId = response.getPayload().getShipmentId();
            storeShipmentInformation(input.getOrderId(), shipmentId);

            return response.getPayload().getLabel();
        } catch (Exception e) {
            throw new InternalError("Calling Merchant Fulfillment API failed", e);
        }
    }

    private void storeShipmentInformation(String orderId, String shipmentId) {
        Map<String, AttributeValue> item = new HashMap<>();
        item.put(SHIPMENTS_TABLE_HASH_KEY_NAME, AttributeValue.fromS(orderId));
        item.put(SHIPMENTS_TABLE_SHIPMENT_ID_ATTRIBUTE_NAME, AttributeValue.fromS(shipmentId));

        PutItemRequest putItemRequest = PutItemRequest.builder()
                .tableName(System.getenv(SHIPMENTS_TABLE_NAME_ENV_VARIABLE))
                .item(item)
                .build();

        DynamoDbClient dynamoDB = DynamoDbClient.builder().build();
        dynamoDB.putItem(putItemRequest);
    }
}

File: use-cases/merchant-fulfillment/code/python/src/create_shipment_handler.py
import os
import json
import boto3
import logging

from src.utils import constants
from src.utils import mfn_utils
from src.utils.api_utils import ApiUtils

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def lambda_handler(event, context):
    # Extract refresh token from event
    refresh_token = event[constants.STATE_MACHINE_REFRESH_TOKEN_KEY_NAME]

    # Extract region code from event
    region_code = event[constants.STATE_MACHINE_REGION_CODE_KEY_NAME]

    # Extract order ID from event
    order_id = event[constants.STATE_MACHINE_ORDER_ID_KEY_NAME]

    logger.info(f"CreateShipment Lambda input: {event}")

    try:
        # Create an instance of the ApiUtils class
        api_utils = ApiUtils(refresh_token, region_code, constants.MFN_API_TYPE)

        # Create shipment for the selected shipping service
        create_shipment_request = get_create_shipment_request_body(payload=event[constants.MFN_ORDER_KEY_NAME], order_id=order_id)

        logger.info(f"Merchant Fulfillment API - CreateShipment request: {create_shipment_request}")

        create_shipment_result = api_utils.call_mfn_api('create_shipment', body=create_shipment_request)

        logger.info(f"Merchant Fulfillment API - CreateShipment response: {create_shipment_result}")

        # Store ShipmentId in DynamoDB
        # Update this section to match your product's logic
        shipment_id = create_shipment_result.payload.shipment_id
        store_shipment_information(order_id, shipment_id)

        # Generating Label format
        label = create_shipment_result.payload.label
        result = {
            constants.LABEL_FORMAT_KEY_NAME: label.label_format,
            constants.LABEL_DIMENSIONS_KEY_NAME: label.dimensions.to_dict(),
            constants.LABEL_FILE_CONTENTS_KEY_NAME: label.file_contents.to_dict()
        }

        return result

    except Exception as e:
        raise Exception("Calling Merchant Fulfillment API failed", e)


def store_shipment_information(order_id, shipment_id):
    item = {
        constants.SHIPMENTS_TABLE_HASH_KEY_NAME: {'S': order_id},
        constants.SHIPMENTS_TABLE_SHIPMENT_ID_ATTRIBUTE_NAME: {'S': shipment_id}
    }

    put_item_request = {
        'TableName': os.environ.get(constants.SHIPMENTS_TABLE_NAME_ENV_VARIABLE),
        'Item': item
    }

    dynamodb = boto3.client(constants.AWS_DYNAMO_DB_CLIENT_KEY_NAME)
    dynamodb.put_item(**put_item_request)


def get_create_shipment_request_body(payload, order_id):
    order_item_list = json.loads(mfn_utils.transform_keys_to_uppercase_first_letter(json.dumps([item for item in payload["orderItems"]])))
    ship_from_address = json.loads(mfn_utils.snake_to_pascal_case(json.dumps(payload["shipFromAddress"])))
    package_dimensions = json.loads(mfn_utils.snake_to_pascal_case(json.dumps(payload["packageDimensions"])))
    weight = json.loads(mfn_utils.snake_to_pascal_case(json.dumps(payload["weight"])))

    shipping_service_options = payload["preferredShippingService"]["ShippingServiceOptions"]
    shipping_service_id = payload["preferredShippingService"]["ShippingServiceId"]
    shipping_service_offer_id = payload["preferredShippingService"]["ShippingServiceOfferId"]

    request = {
        "ShipmentRequestDetails": {
            "AmazonOrderId": order_id,
            "ItemList": order_item_list,
            "ShipFromAddress": ship_from_address,
            "PackageDimensions": package_dimensions,
            "Weight": weight,
            "ShippingServiceOptions": shipping_service_options
        },
        "ShippingServiceId": shipping_service_id,
        "ShippingServiceOfferId": shipping_service_offer_id
    }

    return json.dumps(request)
